README.md
# 東京タワーチケット予約システムジョブアプリケーション

[![CircleCI](https://circleci.com/gh/motionpicture/ttts-jobs.svg?style=svg)](https://circleci.com/gh/motionpicture/ttts-jobs)

## Table of contents

* [Usage](#usage)
* [License](#license)

## Usage

### Environment variables

| Name                                   | Required | Value       | Purpose                                         |
| -------------------------------------- | -------- | ----------- | ----------------------------------------------- |
| `DEBUG`                                | false    | ttts-jobs:* | Debug                                           |
| `NODE_ENV`                             | true     |             | Environment Name                                |
| `SENDGRID_API_KEY`                     | true     |             | SendGrid API Keyf                               |
| `REDIS_HOST`                           | true     |             | Redis Connection                                |
| `REDIS_PORT`                           | true     |             | Redis Connection                                |
| `REDIS_KEY`                            | true     |             | Redis Connection                                |
| `MONGOLAB_URI`                         | true     |             | MongoDB Connection                              |
| `GMO_ENDPOINT`                         | true     |             | GMO Endpoint                                    |
| `GMO_SITE_ID`                          | true     |             | GMO サイトID                                    |
| `EVENT_AGGREGATION_EXPIRES_IN_SECONDS` | true     |             | イベント集計保持期間                            |
| `MAXIMUM_ATTENDEE_CAPACITY`            | true     |             | イベント最大収容人数                            |
| `COGNITO_USER_POOL_ID`                 | true     |             | 入場アカウントを管理するCognitoユーザープールID |
| `AWS_ACCESS_KEY_ID`                    | true     |             | AWSリソースアクセスキー                         |
| `AWS_SECRET_ACCESS_KEY`                | true     |             | AWSリソースアクセスシークレット                 |
| `PROJECT_ID`                           | true     |             | Project ID                                      |
| `CHEVRE_AUTHORIZE_SERVER_DOMAIN`       | true     |             | Chevre API Settings                             |
| `CHEVRE_CLIENT_ID`                     | true     |             | Chevre API Settings                             |
| `CHEVRE_CLIENT_SECRET`                 | true     |             | Chevre API Settings                             |
| `CHEVRE_API_ENDPOINT`                  | true     |             | Chevre API Settings                             |

## License

ISC


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## Unreleased

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## v19.0.0 - 2019-07-19

### Removed

- 継続的ジョブを削除

## v18.0.0 - 2019-07-18

### Changed

- 予約番号と購入番号を分離

### Removed

- 注文照会キーを削除

## v17.0.1 - 2019-07-15

### Changed

- 注文コレクションのインデックス調整

## v17.0.0 - 2019-07-12

### Changed

- Chevreイベント管理バージョンを追加
- Chevreからの作品コードの引き継ぎを調整

### Removed

- パフォーマンスインターフェースから券種情報を削除
- Chevreを使用しないパフォーマンス管理を削除

## v16.0.0 - 2019-07-05

### Removed

- 券種マスタデータから不要な属性(_id,ttts_extension)を削除

## v15.0.0 - 2019-07-04

### Changed

- DBのパフォーマンスデータをChevre化

## v14.0.0 - 2019-07-03

### Changed

- パフォーマンスを開始日時で検索時にstartDateを参照するように調整

### Removed

- 注文取引結果からeventReservationsを削除(order.acceptedOffersへ移行)
- 仮予約インターフェースから座席グレードを削除
- 仮予約インターフェースからrate_limit_unit_in_secondsを削除
- パフォーマンスインターフェースから非推奨属性を削除

## v13.0.0 - 2019-07-01

### Changed

- 予約の入場以外の全属性をChevre化
- 券種インターフェースをChevre化

## v12.0.0 - 2019-06-27

### Removed

- 予約インターフェースから不要な属性を削除
- 予約コレクションからqr_strに対するユニークインデックスを削除

## v11.0.0 - 2019-06-25

### Removed

- 券種からキャンセルチャージ属性を削除
- 予約インターフェースから不要な属性を削除
- 予約検索条件インターフェースから不要な属性を削除

## v10.0.0 - 2019-06-22

### Removed

- 券種からキャンセルチャージ属性を削除

## v9.3.0 - 2019-06-18

### Changed

- パフォーマンスインターフェースを拡張

## v9.2.1 - 2019-06-14

### Fixed

- 売上レポート対象から車椅子余分確保分予約を除外

## v9.2.0 - 2019-06-14

### Changed

- 車椅子の余分確保分を予約として作成するように調整

### Removed

- 在庫インターフェースを削除

## v9.1.0 - 2019-06-11

### Added

- 不要なイベント集計を削除するタスクを追加

### Changed

- 券種のchargeをpriceSpecificationに置換

## v9.0.0 - 2019-06-10

### Added

- イベント予約集計タスクを追加

### Changed

- パフォーマンスコレクションからのリレーション削除対応に合わせて、自動生成パフォーマンスの型を調整
- jsdoc削除
- 予約リポジトリのmongooseモデルをprivate化
- パフォーマンスリポジトリのmongooseモデルをprivate化
- 新しいRedis在庫リポジトリで再構築
- パフォーマンスインターフェースをChevre化に向けて補強
- スクリーンに車椅子タイプ座席をひとつ持つように設計調整

### Removed

- 劇場、スクリーン、作品、券種、券種グループコレクションを削除(一時的に動的データ管理を不可能に変更)
- GMO通知コレクションを削除
- SendGridイベントコレクションを削除
- 在庫作成ジョブを削除
- パフォーマンスデータから不要な項目を削除
- 一時イベントオファーリポジトリを削除
- パフォーマンス残席数更新ジョブを削除
- オファーごとのパフォーマンス残席数更新ジョブを削除
- パフォーマンス集計ジョブを削除

## v8.1.2 - 2019-04-19

### Changed

- 取引コレクションのインデックス調整
- 予約インターフェースをChevreを継承するように拡張
- 組織コレクションのインデックス調整
- 注文取引の購入者プロフィール情報の持ち方をCinerinoに統一
- install mongoose@5.5.1
- install @motionpicture/ttts-domain@13.5.0

## v8.1.1 - 2019-04-04

### Changed

- 返品レポート作成タスク追加のタイミングを、返品取引確定後から、返品処理完了後に変更

## v8.1.0 - 2018-12-23

### Added

- TriggerWebhookジョブを追加

## v8.0.0 - 2018-11-28

### Removed

- 定期的な売上集計ジョブを削除

## v7.4.0 - 2018-11-28

### Added

- 注文取引レポート作成タスクを追加。
- 注文返品取引レポート作成タスクを追加。

## v7.3.3 - 2018-08-08

### Changed

- データ集計を引数で実行できるように。

## v7.3.1 - 2018-06-25

### Added

- 売上レポート用データ集計処理追加。

## v7.2.8 - 2018-03-22

### Fixed

- 販売者都合での返金完了メールの内容調整。

## v7.2.7 - 2018-03-14

### Changed

- イベントの販売情報ごとの在庫状況集計期間を調整。

## v7.2.6 - 2018-03-05
### Changed
- パフォーマンス在庫状況更新処理とパフォーマンス集計データ作成ジョブの対象期間を、必要分に対してやや冗長に設定するように変更。
- イベント販売情報更新ジョブにイベント開始日時fromとthroughを指定できるように変更。

### Fixed
- パフォーマンスの集計データ作成時の、販売情報ごとの予約数カウントロジックを修正。
- 車椅子券種在庫数算出時に、同伴者分の在庫を考慮するように変更。

## v7.2.5 - 2018-02-14
### Changed
- 予約スキーマに注文取引主体属性を追加。
- パフォーマンス集計対象を、日本時間で集計タイミングの日の0時から指定期間分にするように調整。

### Fixed
- パフォーマンス集計が毎秒の仕様であるのに対して、処理時間が数分、という状況に対して、処理自体をチューニング。

## v7.2.4 - 2018-02-02
### Fixed
- 券種マスターデータを修正。
- クレジットカード決済のない注文の返品処理が完了しないバグを修正。

## v7.2.3 - 2018-01-19
### Changed
- パフォーマンスと在庫を91日後から1日分、毎時で常に作成トライするように変更。

## v7.2.2 - 2018-01-18
### Fixed
- 取引に対するタスクがエクスポートされないバグ対応。

## v7.2.1 - 2018-01-17
### Changed
- install latest ttts-domain.

## v7.2.0 - 2018-01-16
### Added
- パフォーマンス在庫状況更新期間を環境変数として追加。

### Changed
- 予約の属性に在庫情報リストを追加。
- イベントと在庫作成ジョブのスケジュール調整。

## v7.1.0 - 2018-01-11
### Added
- 展示イベントの販売情報更新ジョブを追加。

## v7.0.2 - 2018-01-10
### Fixed
- 在庫作成時にスクリーンが存在しない場合に対応。

## v7.0.1 - 2018-01-10
### Added
- 本番環境のjsonマスターデータを追加。

### Removed
- 所有者jsonマスターデータを削除。

## v7.0.0 - 2018-01-10
### Added
- パフォーマンスに対する返品タスク監視ジョブを追加。
- パフォーマンスに対する供給情報ごとの在庫状況を更新するジョブを追加。
- 入場ゲート情報同期ジョブを追加。
- パフォーマンスごとの集計ジョブを追加。
- マスターインポートジョブを追加。
- 注文作成ジョブを追加。

### Removed
- 所有者コレクションを削除(ユーザー情報をCognitoへ移行)

## v6.0.0 - 2017-12-13
### Changed
- 取引と在庫スキーマを追加。予約を在庫を明確に分離。

### Removed
- 不要ジョブやコードを削除。


common/Util/Util.js
"use strict";
/**
 * 共通ユーティリティ
 *
 * @namespace Util
 */
Object.defineProperty(exports, "__esModule", { value: true });
const crypto = require("crypto");
const log4js = require("log4js");
/**
 * 予約プロセス用のロガーを設定する
 * 1ログファイル per 1購入番号
 *
 * @param {string} paymentNo 購入番号
 * @memberOf Util
 */
// tslint:disable-next-line:variable-name
function getReservationLogger(_paymentNo, cb) {
    // todo 購入ログの出力方法を、ローカルファイルロガー以外で新たに考える
    cb(null, log4js.getLogger('system'));
    // let env = process.env.NODE_ENV || 'development';
    // let logDir = `${__dirname}/../../../logs/${env}/reservations/${paymentNo.substr(-1)}`;
    // fs.mkdirs(logDir, (err) => {
    //     if (err) return cb(err, null);
    //     log4js.configure({
    //         appenders: [
    //             {
    //                 category: 'reservation',
    //                 type: 'file',
    //                 filename: `${logDir}/${paymentNo}.log`,
    //                 pattern: '-yyyy-MM-dd'
    //             },
    //             {
    //                 type: 'console'
    //             }
    //         ],
    //         levels: {
    //             reserve: 'ALL'
    //         },
    //         replaceConsole: true
    //     });
    //     cb(null, log4js.getLogger('reservation'));
    // });
}
exports.getReservationLogger = getReservationLogger;
/**
 * ハッシュ値を作成する
 *
 * @param {string} password
 * @param {string} salt
 * @memberOf Util
 */
function createHash(password, salt) {
    const sha512 = crypto.createHash('sha512');
    sha512.update(salt + password, 'utf8');
    return sha512.digest('hex');
}
exports.createHash = createHash;
/**
 * 全角→半角変換
 *
 * @memberOf Util
 */
function toHalfWidth(str) {
    return str.split('').map((value) => {
        // 全角であれば変換
        // tslint:disable-next-line:no-magic-numbers
        return value.replace(/[！-～]/g, String.fromCharCode(value.charCodeAt(0) - 0xFEE0)).replace('　', ' ');
    }).join('');
}
exports.toHalfWidth = toHalfWidth;
/**
 * 半角→全角変換
 *
 * @memberOf Util
 */
function toFullWidth(str) {
    return str.split('').map((value) => {
        // 半角であれば変換
        // tslint:disable-next-line:no-magic-numbers
        return value.replace(/[!-~]/g, String.fromCharCode(value.charCodeAt(0) + 0xFEE0)).replace(' ', '　');
    }).join('');
}
exports.toFullWidth = toFullWidth;
/**
 * 都道府県リスト
 *
 * @memberOf Util
 */
function getPrefectrues() {
    return [
        { code: '01', name: { ja: '北海道', en: 'Hokkaido Government' } },
        { code: '02', name: { ja: '青森県', en: 'Aomori Prefectural Government' } },
        { code: '03', name: { ja: '岩手県', en: 'Iwate Prefectural Government' } },
        { code: '04', name: { ja: '宮城県', en: 'Miyagi Prefectural Government' } },
        { code: '05', name: { ja: '秋田県', en: 'Akita Prefecture' } },
        { code: '06', name: { ja: '山形県', en: 'Yamagata Prefecture' } },
        { code: '07', name: { ja: '福島県', en: 'Fukushima Prefecture' } },
        { code: '08', name: { ja: '茨城県', en: 'Ibaraki Prefecture Government' } },
        { code: '09', name: { ja: '栃木県', en: 'Tochigi Prefecture' } },
        { code: '10', name: { ja: '群馬県', en: 'Gunma Prefecture' } },
        { code: '11', name: { ja: '埼玉県', en: 'Saitama Prefectural Government' } },
        { code: '12', name: { ja: '千葉県', en: 'Chiba Prefectural Government' } },
        { code: '13', name: { ja: '東京都', en: 'Tokyo Metropolitan Government' } },
        { code: '14', name: { ja: '神奈川県', en: 'Kanagawa Prefecture' } },
        { code: '15', name: { ja: '新潟県', en: 'Niigata Prefecture' } },
        { code: '16', name: { ja: '富山県', en: 'Toyama Prefecture' } },
        { code: '17', name: { ja: '石川県', en: 'Ishikawa Prefecture' } },
        { code: '18', name: { ja: '福井県', en: 'Fukui Prefectural Government' } },
        { code: '19', name: { ja: '山梨県', en: 'Yamanashi Prefecture' } },
        { code: '20', name: { ja: '長野県', en: 'Nagano Prefecture' } },
        { code: '21', name: { ja: '岐阜県', en: 'Gifu Prefectural Government' } },
        { code: '22', name: { ja: '静岡県', en: 'Shizuoka Prefecture' } },
        { code: '23', name: { ja: '愛知県', en: 'Aichi Prefecture' } },
        { code: '24', name: { ja: '三重県', en: 'Mie Prefecture' } },
        { code: '25', name: { ja: '滋賀県', en: 'Shiga Prefecture' } },
        { code: '26', name: { ja: '京都府', en: 'Kyoto Prefecture' } },
        { code: '27', name: { ja: '大阪府', en: 'Osaka Prefectural Government' } },
        { code: '28', name: { ja: '兵庫県', en: 'Hyogo Prefecture' } },
        { code: '29', name: { ja: '奈良県', en: 'Nara Prefecture' } },
        { code: '30', name: { ja: '和歌山県', en: 'Wakayama Prefecture' } },
        { code: '31', name: { ja: '鳥取県', en: 'Tottori Prefecture' } },
        { code: '32', name: { ja: '島根県', en: 'Shimane Prefectural Government' } },
        { code: '33', name: { ja: '岡山県', en: 'Okayama Prefecture' } },
        { code: '34', name: { ja: '広島県', en: 'Hiroshima Prefecture' } },
        { code: '35', name: { ja: '山口県', en: 'Yamaguchi Prefecture' } },
        { code: '36', name: { ja: '徳島県', en: 'Tokushima Prefecture' } },
        { code: '37', name: { ja: '香川県', en: 'Kagawa Prefectural Government' } },
        { code: '38', name: { ja: '愛媛県', en: 'Ehime Prefectural Government' } },
        { code: '39', name: { ja: '高知県', en: 'Kochi Prefecture' } },
        { code: '40', name: { ja: '福岡県', en: 'Fukuoka Prefecture' } },
        { code: '41', name: { ja: '佐賀県', en: 'Saga Prefectural Government' } },
        { code: '42', name: { ja: '長崎県', en: 'Nagasaki Prefecture' } },
        { code: '43', name: { ja: '熊本県', en: 'Kumamoto Prefecture' } },
        { code: '44', name: { ja: '大分県', en: 'Oita Prefecture' } },
        { code: '45', name: { ja: '宮崎県', en: 'Miyazaki Prefecture' } },
        { code: '46', name: { ja: '鹿児島県', en: 'Kagoshima Prefecture' } },
        { code: '47', name: { ja: '沖縄県', en: 'Okinawa Prefecture' } }
    ];
}
exports.getPrefectrues = getPrefectrues;
